### PR TITLE
Block parser + Binary expression parser

### PR DESCRIPTION
**Pull request by kvverti**
This pull request depends on #8 . Adds a Block struct, ExpressionSt struct, and Statement enum. Parsers:
```
Block ::= "{" (Statement)* [Expression] "}"
ExpressionSt ::= Expression ";"
```

Also re-added the binary expression parser, as the expression parser is required for implementing blocks. Binary expressions are left associative.

### CODE REVIEW

**Comment by kvverti** [comment]
Vote :+1: for dropping the semicolon on block statements, or :-1: for requiring the semicolon on block statements. e.g.
```
{
  { 3 + 2 }; // this semicolon
  true
}
```

File: wright/src/grammar/parsers/expression/binary_expression.rs [hunk 1]
```diff
@@ -13,9 +65,86 @@ impl<'s> BinaryExpression<'s> {
...
+            Self::cmp,
+            &[BinaryOp::EqEq, BinaryOp::NotEq],
+        )(input)
+    }
+
     /// Parse a binary expression in source code.
-    pub fn parse(input: Fragment<'s>) -> IResult<Fragment<'s>, Self> {
-        todo!("binary expression parser")
+    pub fn parse(input: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        Self::eq(input)
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/expression/binary_expression.rs]: Binary expression parsing should return a `BinaryExpression`, not an `Expression`.

File: wright/src/grammar/parsers/expression/binary_expression.rs [hunk 2]
```diff
@@ -1,8 +1,60 @@
...
+        let mut res = Err(nom::Err::Error((input, ErrorKind::Alt)));
+        for op in ops {
+            res = value(*op, tag(op.token()))(input);
+            if res.is_ok() {
+                break;
+            }
+        }
+        res
+    }
+}
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/expression/binary_expression.rs]: this function always returns an error when the list is empty? I feel like there perhaps is a better way to do this.

File: wright/src/grammar/parsers/expression.rs [hunk 3]
```diff
@@ -10,31 +10,57 @@ pub(crate) mod binary_expression;
...
+    pub fn primary(input: Fragment<'s>) -> IResult<Fragment<'s>, Self> {
+        alt((
+            map(Block::parse, Expression::Block),
+            map(Parens::parse, Expression::Parens),
+            map(NumLit::parse, Expression::NumLit),
+            map(CharLit::parse, Expression::CharLit),
+            map(StringLit::parse, Expression::StringLit),
+            map(BooleanLit::parse, Expression::BooleanLit),
+            map(Underscore::parse, Expression::Underscore),
+            map(Identifier::parse, Expression::Identifier),
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/expression.rs]: I think I will replace this by adding to the `ToExpression` trait.

File: wright/src/grammar/parsers/expression/block.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                    preceded(
+                        multispace0,
+                        Statement::parse,
+                    ),
+                ),
+                delimited(
+                    multispace0,
+                    opt(Expression::parse),
+                    multispace0,
+                )
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/expression/block.rs]: Perhaps swap opt and delimited here?

File: wright/src/grammar/parsers/expression/block.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            |parse| {
+                let (statements, expr) = Self::inner(parse).unwrap().1;
+                Block {
+                    frag: parse,
+                    statements,
+                    result: expr.map(Box::new),
+                }
+            },
+        )(input)
+    }
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/expression/block.rs]: rewrite using new `with_input` function

File: wright/src/grammar/parsers/statement/expression_statement.rs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+use nom::sequence::{preceded, terminated};
+
+impl<'s> ExpressionSt<'s> {
+    fn inner(frag: Fragment<'s>) -> IResult<Fragment<'s>, Expression<'s>> {
+        terminated(
+            Expression::parse,
+            preceded(
+                multispace0,
+                ch(';'),
+            ),
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/statement/expression_statement.rs]: use `delimited` insead of `preceded`. We should also parse through following white space

File: wright/src/grammar/parsers/statement/expression_statement.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    pub fn parse(frag: Fragment<'s>) -> IResult<Fragment<'s>, Self> {
+        map(
+            recognize(Self::inner),
+            |parse| {
+                ExpressionSt {
+                    frag: parse,
+                    inner: Self::inner(parse).unwrap().1,
+                }
+            }
+        )(frag)
```
Comment (Alfriadox) [inline wright/src/grammar/parsers/statement/expression_statement.rs]: rewrite using new `with_input` function.

File: wright/src/grammar/parsers/expression/binary_expression.rs (hunk 2 above)
Comment (kvverti) [reply wright/src/grammar/parsers/expression/binary_expression.rs]: Maybe the function can take `(first: BinaryOp, rest: &[BinaryOp])` and then have a macro for varargs calling.

File: wright/src/grammar/parsers/statement/expression_statement.rs (hunk 6 above)
Comment (kvverti) [reply wright/src/grammar/parsers/statement/expression_statement.rs]: Do we want to write parsers as 
```
<whitespace> <subobj> <subobj> <whitespace>
```
or as
```
<subobj> <whitespace> <subobj>
```
If we stay consistent we can pick either one.

File: wright/src/grammar/parsers/statement/expression_statement.rs (hunk 7 above)
Comment (kvverti) [reply wright/src/grammar/parsers/statement/expression_statement.rs]: Made `Block::parse` and `binary_parser` use `with_input`

File: wright/src/grammar/parsers/expression/block.rs (hunk 4 above)
Comment (kvverti) [reply wright/src/grammar/parsers/expression/block.rs]: Pushed reformulation

File: wright/src/grammar/parsers/expression/binary_expression.rs (hunk 1 above)
Comment (kvverti) [reply wright/src/grammar/parsers/expression/binary_expression.rs]: Ready for second review

**Comment by kvverti** [comment]
Closed due to subsumption.

**Pull request closed without merging**